javalanglib/src/main/scala/java/lang/Float.scala

```diff
@@ -72,8 +76,229 @@ object Float {
+   *
+   *  ---
+   *
+   *  As proven in the paper "How to Read Float Point Numbers Accurately" by
+   *  William D. Clinger, there is no solution that does not require big
+   *  integer arithmetic at some point. We take inspiration from the
+   *  `AlgorithmR` from that paper, which takes an initial value "close" to the
+   *  best approximation and improves it by 1 ULP. Since we already have a
+   *  close approximation (one that is at most 1 ULP away from the best one),
+   *  we can use that. However, we can dramatically simplify the algorithm
+   *  because we can leverage Double arithmetics to parse only a Float. In
+   *  particular, we can accurately compute and represent the two adjacent
+   *  Floats that enclose the best approximation, as well as the midpoint
+   *  between those, which is a Double. We receive those from
+   *  `parseFloatDecimal`, which already had to compute them in order to decide
+   *  whether a correction was needed. The only real thing we keep from the
+   *  paper is the step 3: how to accurately compare that midpoint with the
+   *  exact value represented by the string, using big integer arithmetics.
+   *  This allows us to decide whether we need to round up, down, or break a
+   *  tie to even.
+   *
+   *  `AlgorithmR` in the paper is generic wrt. the bases of the input and
+   *  output. In our case, the input base Δ is 10 and the output base β is 2.
+   */
+  private def parseFloatDecimalCorrection(integralPartStr: String,
+      fractionalPartStr: String, exponentStr: String,
+      zDown: scala.Float, zUp: scala.Float, mid: scala.Double): scala.Float = {
+
+    // 1. Accurately parse the string with the representation f × 10ᵉ
+
+    val f: BigInteger = new BigInteger(integralPartStr + fractionalPartStr)
+    val e: Int = Integer.parseInt(exponentStr) - fractionalPartStr.length()
+
+    /* Note: we know that `e` is "reasonable". If it were way too big or way
+     * too small, the original JS `parseFloat` in `parseFloatDecimal` would
+     * have returned `Infinity` or `0.0`, respectively. In that case, we would
+     * have selected the first branch, and never called
+     * `parseFloatDecimalCorrection`.
+     */
+
+    // 2. Accurately decompose `mid` with the representation m × 2ᵏ
+
+    val mbits = 52 // number of bits of the mantissa (without the implicit '1')
+    val kbits = 11 // number of bits of the exponent
+    val bias = (1 << (kbits - 1)) - 1 // the bias of the exponent
+
+    val midBits = Double.doubleToLongBits(mid)
+    val biasedK = (midBits >> mbits).toInt
+
+    /* Because `mid` is a double value halfway between two floats, it cannot
+     * be a double subnormal (even if the two floats that surround it are
+     * subnormal floats).
+     */
+    if (biasedK == 0)
+      throw new AssertionError("parseFloatCorrection was given a subnormal mid: " + mid)
+
+    val m = BigInteger.valueOf((midBits & ((1L << mbits) - 1)) | (1L << mbits))
```

Comment:
`(1L << (mbits + 1)) - 1`? Feels easier to understand.

Reply:
That would not be equivalent, unless I'm mistaken.

I've decomposed the computation into `mExplicitBits` and `mImplicit1Bit`, which makes things clearer.